File: LLWallPaper.App/Stores/FavoritesStore.cs
using System.IO;
using System.Text.Json;
using LLWallPaper.App.Models;
using LLWallPaper.App.Utils;

namespace LLWallPaper.App.Stores;

public sealed class FavoritesStore
{
    private readonly AppLogger _logger;
    private Favorites _state = new();

    public FavoritesStore(AppLogger logger)
    {
        _logger = logger;
    }

    public void Load()
    {
        AppPaths.EnsureDirectories();
        if (!File.Exists(AppPaths.FavoritesPath))
        {
            Save();
            return;
        }

        try
        {
            var json = File.ReadAllText(AppPaths.FavoritesPath);
            _state =
                JsonSerializer.Deserialize<Favorites>(json, JsonOptions.Default) ?? new Favorites();
        }
        catch (Exception ex)
        {
            _logger.Error("Failed to load favorites, resetting.", ex);
            _state = new Favorites();
        }
    }

    public bool IsFavorite(string key) => _state.FavoriteKeys.Contains(key);

    public bool IsBlocked(string key) => _state.BlockedKeys.Contains(key);

    public IReadOnlyCollection<string> FavoriteKeys => _state.FavoriteKeys;

    public IReadOnlyCollection<string> BlockedKeys => _state.BlockedKeys;

    public void ToggleFavorite(string key)
    {
        if (_state.FavoriteKeys.Contains(key))
        {
            _state.FavoriteKeys.Remove(key);
        }
        else
        {
            _state.FavoriteKeys.Add(key);
        }

        Save();
    }

    public void ToggleBlocked(string key)
    {
        if (_state.BlockedKeys.Contains(key))
        {
            _state.BlockedKeys.Remove(key);
        }
        else
        {
            _state.BlockedKeys.Add(key);
        }

        Save();
    }

    private void Save()
    {
        AppPaths.EnsureDirectories();
        var json = JsonSerializer.Serialize(_state, JsonOptions.Default);
        File.WriteAllText(AppPaths.FavoritesPath, json);
    }
}

File: LLWallPaper.Mac/Sources/LLWallPaperMacCore/Infrastructure/FavoritesStore.swift
import Foundation

public final class FavoritesStore: FavoritesReading {
  private let logger: AppLogger
  private var state = Favorites()

  public init(logger: AppLogger) {
    self.logger = logger
  }

  public var favoriteKeys: Set<String> {
    Set(state.favoriteKeys)
  }

  public var blockedKeys: Set<String> {
    Set(state.blockedKeys)
  }

  public func load() {
    do {
      try AppPaths.ensureDirectories()
      guard FileManager.default.fileExists(atPath: AppPaths.favoritesUrl.path) else {
        save()
        return
      }

      let data = try Data(contentsOf: AppPaths.favoritesUrl)
      state = try JsonCoding.decoder.decode(Favorites.self, from: data)
    } catch {
      logger.error("Failed to load favorites, resetting.", error)
      state = Favorites()
    }
  }

  public func isFavorite(_ key: String) -> Bool {
    state.favoriteKeys.contains(key)
  }

  public func isBlocked(_ key: String) -> Bool {
    state.blockedKeys.contains(key)
  }

  public func toggleFavorite(_ key: String) {
    toggle(key: key, in: \.favoriteKeys)
    save()
  }

  public func toggleBlocked(_ key: String) {
    toggle(key: key, in: \.blockedKeys)
    save()
  }

  private func toggle(key: String, in keyPath: WritableKeyPath<Favorites, [String]>) {
    if let index = state[keyPath: keyPath].firstIndex(of: key) {
      state[keyPath: keyPath].remove(at: index)
    } else {
      state[keyPath: keyPath].append(key)
    }
  }

  private func save() {
    do {
      try AppPaths.ensureDirectories()
      let data = try JsonCoding.encoder.encode(state)
      try data.write(to: AppPaths.favoritesUrl, options: .atomic)
    } catch {
      logger.error("Failed to save favorites.", error)
    }
  }
}
